Fix core sporadic boot on Intel ADSP cAVS 2.5 platforms

lyakh opened the pull request:
Only for discussion so far

dcpleung committed:
cavs: disable timer interrupt on cpu power off

Since the timer interrupt is delivered to all cores at the same
time, it needs to be disabled (or masked) to prevent it from
waking a suspended CPU core.

Note that there is no need to re-enable (unmask) timer
interrupt separately as it is being taken care of in
smp_timer_init() when the CPU re-inits itself.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

kv2019i commented on soc/xtensa/intel_adsp/common/include/soc.h:
```diff
@@ -135,4 +135,12 @@ static inline bool intel_adsp_ptr_is_sane(uint32_t sp)
+	__asm__ volatile("rsr %0, PRID" : "=r"(prid));
+	return prid;
+}
+
```
@nashif beat you to it -> https://github.com/zephyrproject-rtos/zephyr/pull/46659

kv2019i reviewed (changes requested):
Excellent @lyakh ! Minor nit as Anas already made a patch to make prid more easily available.

dcpleung commented:
Not related to this but something to consider is that the code is using the same stack space for all cores. So if two cores are waking up at the same time, they are going to step on each other and overwriting the stack.

andyross commented on soc/xtensa/intel_adsp/common/soc_mp.c:
```diff
@@ -77,6 +77,9 @@ __imr void z_mp_entry(void)
 {
 	cpu_early_init();
 
+	if (prid() != start_rec.cpu)
+		k_cpu_idle();
```
Experimentally it is for all the intel_adsp devices.  But at the hardware design level it can be anything.  ESP32 had funny numbers in that register.  Worst-case is that we'd have to shift over to _current_cpu->id if this changes.

nashif replied:
if statement needs braces

lyakh commented:
Thinking again why this problem only triggers on one laptop model. What we think is happening here is that
* a core is allowed to enter low-power
* the core calls waiti and enters low-power
* an external trigger (exception, interrupt) wakes up the core and it runs through its reset vector

And that's the difference with the normal waiti entry / exit flow when low-power isn't allowed: then after the trigger the core jumps to the exception / interrupt handling vector.

What I see on all TGL / ADL systems, except for that laptop, that the fix from @dcpleung that disables the timer interrupt is enough. But on that laptop it isn't. Moreover, disabling all l2 and l5 interrupts doesn't help either. And unlike when processing the interrupt normally in waiti, when you enter the interrupt handling code, when waking up via the reset handler I don't see a possibility to find out what was the wakeup cause. So it looks like that laptop has an additional sporadic DSP core wakeup source.

lyakh committed:
xtensa: return to idle() when booting sporadically

To power down secondary cores on cAVS 2.5 platforms the primary core
enables the power-saving mode for the respective secondary core and
waits until that core enters idle() and executes the waiti
instruction at which point the core should enter a lower-power mode.
However, that core can then also automatically wake up and execute
its reset path if, e.g. an interrupt is delivered to it. However, it
isn't entirely clear which events are able to wake up cores from that
state. Disabling interrupts on the interrupt controller didn't seem
to prevent that from happening completely. In particular a specific
ADL notebook seems to be susceptible to this problem. Checking for
such sporadic boots and returning to idle fixes the problem.

BugLink: https://github.com/zephyrproject-rtos/zephyr/issues/46372
BugLink: https://github.com/thesofproject/sof/issues/5733
Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh commented:
interestingly, the problem in https://github.com/thesofproject/sof/issues/5733 is solved with just the first commit, but the smoke test needs both...

kv2019i reviewed (dismissed):
Thanks @lyakh ! Maybe not ideal, but core waking up and causing havoc on the stack is for sure much, much worse.

marc-hb commented:
> Will need to update the commit message on the second commit:

Considering this intermittent failure reproducible only on some specific products took multiple experts months to address, shouldn't more of these explanations be found in the (not product specific) source code itself? I don't understand the code change but it seems to have surprisingly terse and mundane comments compared to the magnitude of the problem.

My 2 cents.

lyakh commented:
CI: compliance is complaining about spaces in inline asm strings, i.e. is a false positive. @nashif the previous version was also approved by @kv2019i but unfortunately that approval got removed when I updated the PR, addressing @dcpleung 's comments

nashif commented:
ok, please fix the PR body as well.

nashif commented:
need @dcpleung  +1
